Draft of retworkx: A High-Performance Graph Library for Python (#455)
Qiskit/retworkx: paper.md

# Related work

To address the performance issues in Qiskit, we explored several graph library alternatives. _SNAP_ [@Leskovec2016], _graphtool_ [@Peixoto2014], and _igraph_ [@Csardi2006] are stable Python libraries written in C or C++ that can replace _NetworkX_.

However, there was a strong desire to keep the flexibility that _NetworkX_ provided for exploring and interacting with the graphs, which precluded custom data structures. The investigated graph libraries either had issues integrating with Qiskit or APIs that were too rigid, such that the migration of existing code was more complex than desired. Thus, the main contribution of _retworkx_ is keeping the ease of use of _NetworkX_ without sacrificing performance.

# Graph data structures

> _retworkx_ provides two core data structures: `PyGraph` and `PyDiGraph`. They correspond to undirected and directed graphs, respectively. Graphs describe a set of nodes and the edges connecting pairs of those nodes. Internally, _retworkx_ leverages the _petgraph_ [@bluss2021] and the _PyO3_ libraries [@Hewitt2021] to store the graphs.

Reviewer: A more accurate statement here might be better. Something like:
```suggestion
_retworkx_ provides two core data structures: `PyGraph` and `PyDiGraph`. They correspond to undirected and directed graphs, respectively. Graphs describe a set of nodes and the edges connecting pairs of those nodes. Internally, _retworkx_ leverages the _petgraph_ library [@bluss2021] that provides the underlying graph data structure and the _PyO3_ library [@Hewitt2021] for the Python bindings.
```

Nodes and edges of the graph may also be associated with weights. Weights can contain arbitrary data, such as node labels or edge lengths. Any Python object can be a weight, which makes the library flexible because no assumptions are made about the weight types. 

> A defining characteristic of _retworkx_ graphs is that each node maps to a non-negative integer node index, and similarly, each edge maps to an edge index. Those indices uniquely determine nodes and edges during the graph object's lifetime. Moreover, the indices provide a clear separation between the underlying graph structure and the data associated with weights.
> 
> _retworkx_ operates on weights with callbacks. Callbacks are functions that take weights and return statically typed data. They resemble the named attributes in _NetworkX_. Callbacks are beneficial because they bridge the arbitrary stored data with the static types _retworkx_ expects. We illustrate indices and callbacks usage with an example.

Reviewer: While reading, it feels more natural for these paragraphs to be in opposite order.

```suggestion
_retworkx_ operates on weights with callbacks. Callbacks are functions that take weights and return statically typed data. They resemble the named attributes in _NetworkX_. Callbacks are beneficial because they bridge the arbitrary stored data with the static types _retworkx_ expects. We illustrate indices and callbacks usage with an example.

A defining characteristic of _retworkx_ graphs is that each node maps to a non-negative integer node index, and similarly, each edge maps to an edge index. Those indices uniquely determine nodes and edges during the graph object's lifetime. Moreover, the indices provide a clear separation between the underlying graph structure and the data associated with weights.
```

A defining characteristic of _retworkx_ graphs is that each node maps to a non-negative integer node index, and similarly, each edge maps to an edge index. Those indices uniquely determine nodes and edges during the graph object's lifetime. Moreover, the indices provide a clear separation between the underlying graph structure and the data associated with weights.

_retworkx_ operates on weights with callbacks. Callbacks are functions that take weights and return statically typed data. They resemble the named attributes in _NetworkX_. Callbacks are beneficial because they bridge the arbitrary stored data with the static types _retworkx_ expects. We illustrate indices and callbacks usage with an example.

## Example

```python
import retworkx

graph = retworkx.PyGraph()

a = graph.add_node("A")
b = graph.add_node("B")
c = graph.add_node("C")

graph.add_edges_from([(a, b, 1.5), (a, c, 5.0), (b, c, 2.5)])
retworkx.dijkstra_shortest_paths(graph, a, c, weight_fn=float)
```

> In the provided example, we find the shortest path from $A$ to $C$ using Dijkstra's algorithm. Firstly, we create the graph and add three nodes labeled A, B, and C. Each time **`add_node`** is called, the method returns the node index associated with the newly added node.

Reviewer 2: Do we need a citation for `Dijkstra's algorithm` here? (I know it's famous and well known but just seems like something missing)

Reviewer 2: ```suggestion
In the provided example, we find the shortest path from $A$ to $C$ using Dijkstra's algorithm [@Dijkstra1959ANO]. Firstly, we create the graph and add three nodes labeled A, B, and C. Each time **`add_node`** is called, the method returns the node index associated with the newly added node.
```

Author: Dijkstra's algorithm is assumed to be common knowledge, but oh well. For example, I cited Rust because it is fairly recent (< 10 years). I will read the 1959 paper before citing it

Author: I am including it, the paper is relevant

Secondly, we add three edges connecting the nodes. **`add_edges_from`** accepts a list containing tuples of node indices and the edge weight, and returns a list of associated edge indices.

Lastly, we compute the shortest path from $A$ to $C$. We provide the graph, the source node, and the target node to **`dijkstra_shortest_paths`**. In addition, we also provide a callback to **`weight_fn`**, which is the built-in **`float`** function for this case. The algorithm returns a mapping containing the shortest path, $A \rightarrow B \rightarrow C$.

# Use Cases

TODO

## Graph Creation

TODO: benchmark using 9th DIMACS challenge dataset [@Demetrescu2016].

## Shortest Path

Talk about wide-range of applications. TODO: benchmark using 9th DIMACS challenge dataset [@Demetrescu2016].

## Graph Isomorphism

> Talk about definition, applications [@Cordella2001] and VF2++ [@Juttner2018]. TODO: benchmark on ICPR2014 [@Carletti2015].

Reviewer 3: I have also found a recent benchmark for graph isomorphism which seems reasonable: [Benchmark Graphs for Practical Graph Isomorphism](https://arxiv.org/abs/1705.03686), with the data for the graphs being [here](https://www.lics.rwth-aachen.de/go/id/rtok/). If you guys find this one useful, I can add to the references.

Reviewer: I have tried this benchmark in the recent past but as expected, it's really difficult and VF2 algorithm is slow. I suggest using an easy and more practical benchmark.

Reviewer 3: That makes sense. Thanks for the clarification.

> Talk about definition, applications [@Cordella2001] and VF2++ [@Juttner2018]. TODO: benchmark on ICPR2014 [@Carletti2015] or on the more challenging [@Neuen2017].

Reviewer 3: ```suggestion
Talk about definition, applications [@Cordella2001], [@Clark2014],  and VF2++ [@Juttner2018]. TODO: benchmark on ICPR2014 [@Carletti2015] or on the more challenging [@Neuen2017].
```

## Maximum Weight Matching

Talk about Blossom algorithm [@Edmonds1965], applications in error correcting codes, cite qtcodes [@Jha2021]. TODO: benchmark on graphs based on Topological Codes [@Criger2018].

# Acknowledgements

> TODO

Reviewer 2: ```suggestion
We thank Kevin Krsulich for his help in getting retworkx ready for use by Qiskit, Lauren Capelluto and Itoko Toshinari for their continued contributing and help with code review, and all the retworkx contributors who have helped retworkx improve over time.
```
